src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorUprooted.java

```diff
@@ -0,0 +1,564 @@
+                    readPos += elementLength;
+                    break;
+                case H:
+                case P:
+                    break;
+                default:
+                    throw new GATKException("Unsupported cigar operator: " + ce.getOperator());
+            }
+        }
+        return snp;
+    }
+
+    protected static int[] calculateIsIndel( final SAMRecord read, final EventType mode ) {
+        final int[] indel = new int[read.getReadBases().length];
+        int readPos = 0;
+        for ( final CigarElement ce : read.getCigar().getCigarElements() ) {
+            final int elementLength = ce.getLength();
+            switch (ce.getOperator()) {
+                case M:
+                case EQ:
+                case X:
+                case S:
+                {
+                    readPos += elementLength;
+                    break;
+                }
+                case D:
+                {
+                    final int index = ( read.getReadNegativeStrandFlag() ? readPos : readPos - 1 );
+                    updateIndel(indel, index, mode, EventType.BASE_DELETION);
+                    break;
+                }
+                case I:
+                {
+                    final boolean forwardStrandRead = !read.getReadNegativeStrandFlag();
+                    if( forwardStrandRead ) {
+                        updateIndel(indel, readPos - 1, mode, EventType.BASE_INSERTION);
+                    }
+                    readPos += elementLength;
+                    if( !forwardStrandRead ) {
+                        updateIndel(indel, readPos, mode, EventType.BASE_INSERTION);
+                    }
+                    break;
+                }
+                case N:
+                case H:
+                case P:
+                    break;
+                default:
+                    throw new GATKException("Unsupported cigar operator: " + ce.getOperator());
+            }
+        }
+        return indel;
+    }
+
+    private static void updateIndel(final int[] indel, final int index, final EventType mode, final EventType requiredMode) {
+        if ( mode == requiredMode && index >= 0 && index < indel.length )
+            // protect ourselves from events at the start or end of the read (1D3M or 3M1D)
+            indel[index] = 1;
+    }
+
+    protected static double[] calculateFractionalErrorArray( final int[] errorArray, final byte[] baqArray ) {
+        if(errorArray.length != baqArray.length ) {
+            throw new GATKException("Array length mismatch detected. Malformed read?");
+        }
+
+        final int BLOCK_START_UNSET = -1;
+
+        final double[] fractionalErrors = new double[baqArray.length];
+        Arrays.fill(fractionalErrors, 0.0);
+        boolean inBlock = false;
+        int blockStartIndex = BLOCK_START_UNSET;
+        int iii;
+        for( iii = 0; iii < fractionalErrors.length; iii++ ) {
```

Comment:
Please add a comment explaining what's happening in this for loop as you walk through the fractionalErrors.